doc: add project layout  definition (#4642)
operator-framework/operator-sdk: website/content/en/docs/overview/project-layout.md

---
title: "Project Layout"
linkTitle: "Project Layout"
weight: 2
description: A description of the layout of projects built with Operator SDK
---

## Operator SDK Project Layout

> All projects built with SDK have a common base structure which follows the Kubebuilder-style for [What’s in a basic project?][kb-whats in-a-basic-project?]. Each Operator language type which is supported by the CLI has its own specific directories and files accordingly to their language.

Reviewer: ```suggestion
All projects initialized with `operator-sdk init` have a common base structure which builds on [kubebuilder's project layout][kb-whats in-a-basic-project?]. Each project type is customized further with code of that type's language.
```

Author: done as suggested 👍

| `config/` | Contains configuration files to launch your project on a cluster. See that plugins might use it to provide their facilities. For example, for the CLI be able to help you out create your operator bundle it will look for the CRD's and CR's which are scaffold in this directory. Also, see that you will find here as well all [Kustomize][Kustomize] YAML definitions. |
| `config/crd/` | Contains the [Custom Resources Definitions][k8s-crd-doc] |
| `config/default/` | contains a [Kustomize base][kustomize-base] for launching the controller in a standard configuration. |
| `config/manager/` | Contains the manifests to launch your operator project as pods on the cluster |
| `config/manifests/` | Contains the base to generate your OLM manifests in the bundle directory |
| `config/prometheus/` | Contains the manifests required to enable project to serve metrics to [Prometheus][kb-metrics] |
| `config/scorecard/` | Contains the manifests required to allow you test your project with [Scorecard][scorecard] |
| `config/rbac/` | Contains the [RBAC][k8s-rbac] permissions required to run your project |
| `config/samples/` | Contains the [Custom Resources][k8s-cr-doc] |

> | `hack/` | Contains the file used to scaffold the lincense in the head of your project files ||

Reviewer 2: ```suggestion
| `hack/` | Contains utility files, e.g. the file used to scaffold the license header for your project files |
```

> | `bundle.Dockerfile` | The project [bundle][bundle] Dockerfile |

Reviewer 2: ```suggestion
| `bundle.Dockerfile` | The Dockerfile to build the [bundle][bundle] image |
```

> ### Ansible

Reviewer 3: A quick question. This is similar to the project layout which is explained here: https://sdk.operatorframework.io/docs/building-operators/ansible/reference/scaffolding/.
would it be better to remove that section and point it here ?

Author: great point varchar. Let's remove the helm one and keep it centralized and then, indeed remove the duplications and address some nits fixes.

> Now, let's check the files and directories which are specific for Ansible.

Reviewer 2: ```suggestion
Now, let's look at the files and directories specific to Ansible-based operators.
```

| File/Directory | Description | 
| ------ | ----- |

> |`config/testing/` | Manifest files that can help you to test your project. For example, to change the configuration of the image policy for your [Molecule tests][ansible-test-guide] tests and to enable the Ansible logs in the debug level. |

Reviewer 2: ```suggestion
|`config/testing/` | Manifest files to help you test your project. For example, to change the image policy for your [Molecule tests][ansible-test-guide] or to enable debug level in the Ansible logs. |
```

|`molecule/` | Contain the manifests for your [Molecule][molecule] tests |
|`molecule/default` | Contains the default [Molecule][molecule] task |
|`molecule/kind` | Contains the [Molecule][molecule] task to be executed on the cluster |
|`playbooks/` | Contains the Ansible playbooks|
|`roles/` | Contains the Ansible role files for each Kind scaffold |
|`requirements.yml` | This file specifies Ansible dependencies that need to be installed for your operator to function. |
|`watches.yaml` | Use to configure the [Ansible watches][ansible-watches]. |

### Golang 

> Now, let's check the files and directories which are specific for Helm.

Reviewer 2: ```suggestion
Now, let's look at the files and directories specific to Golang operators.
```

| File/Directory | Description |
| ------ | ----- |
|`api/` | Contains the api definition |

> |`config/certmanager` |  Contains the Kustomize manifests which configure the cert-manager which are used with the Webhooks |

Reviewer 2: ```suggestion
|`config/certmanager` |  Contains the Kustomize manifests which configure the cert-manager by the Webhooks |
```

|`config/webhook` | Contains the Kustomize manifests to configure the Webhook |
|`controllers` |  Contains the controllers |

> |`main.go` | Its implement the project initialization |

Reviewer 2: ```suggestion
|`main.go` | Implements the project initialization |
```

> ### Helm 

Reviewer 3: same thing for helm. can we remove this to point it here https://sdk.operatorframework.io/docs/building-operators/helm/reference/project_layout/ to avoid duplication?

Author: great point varchar. Let's remove the helm one and keep it centralized.

> Now, let's check the files and directories which are specific for Helm.

Reviewer 2: ```suggestion
Now, let's look at the files and directories specific to Helm-based operators.
```

---
title: "Project Layout"
linkTitle: "Project Layout"
weight: 2
description: >
    What is the layout of the projects build with Operator SDK?
---

## Operator SDK Project Layout

> All projects built with SDK have a common base structure which follows the Kubebuilder-style for [What’s in a basic project?][kb-whats in-a-basic-project?]. Each type has specific directories and specific language files.

Author: .> All projects initialized with `operator-sdk init` have a common base
The files are not only added/changed/ via init
> Each project type is customized further with types, controller logic, etc. implemented in that type's language.
we have no controller logic scaffold at all. 

I'd prefer something simple and generic. However, I agree that it has how to improvements. Pushed a change based on your comment.

> All projects built with SDK have a common base structure which follows the Kubebuilder-style for [What’s in a basic project?][kb-whats in-a-basic-project?]. Each type has specific directories and specific language files.

Author: also, see : https://github.com/operator-framework/operator-sdk/pull/4642#discussion_r594428112

> All projects built with SDK have a common base structure which follows the Kubebuilder-style for [What’s in a basic project?][kb-whats in-a-basic-project?]. Each type has specific directories and specific language files.

Reviewer: >we have no controller logic, etc scaffold for all.

There certainly is, although at different levels of complexity. You can remove that part though if you want to keep this paragraph generic.
